mut/redirects/redirect_main.py

```diff
@@ -0,0 +1,396 @@
+    
+def write_to_file(rules, output):
+    
+    # I'm not sure how we want to handle the file name - maybe it makes sense to have it as a parameter?
+    file = open(output + '/htaccess_test.txt', 'w')
+    
+    for rule in rules:
+        line = 'Redirect '
+        
+        if rule.is_temp:
+            line += '302 '
+        else:
+            line += '301 '
+            
+        line += str(rule.old_url) + ' ' + str(rule.new_url)
+        
+        file.write(line)
+        file.write('\n')
+        
+    file.close()
+
+def parse_source_file(source_path, output):
+    
+    line_num = 0
+    
+    with open(source_path) as file:
+        for line in file:
+            
+            line_num += 1
+            
+            # strip \n from line
+            line = line.strip()
+            
+            # regex to see if we are dealing with a keyword - define, symlink, or raw
+            if re.search('(define|symlink|raw)', line):
+                keyword_split = line.split(':', 1)
+                
+                # define:
+                if keyword_split[0] == 'define':
+                    type_split = keyword_split[1].split(' ')
+                    
+                    key = type_split[1]
+                    
+                    
+                    if len(type_split) > 3:
+                        
+                        value = []
+                        
+                        for x in range (2, len(type_split)):
+                            value.append(type_split[x])
+                            
+                    else:
+                        value = type_split[2]
+                        
+                    definition = [key, value]
+                    
+                    global definitions
+                    definitions.append(definition)
+                    
+                    # see if any definitions have to be substituted with other definitions:
+                    for x in range(0, len(definitions)):
+                        sub_regex = '{(.*?)}'
+                        
+                        if type(definitions[x][1]) is str:
+                            matches = re.findall(sub_regex, definitions[x][1], re.DOTALL)
+    
+                            if matches:
+                                # loop through each match
+                                for y in range(0, len(matches)): 
+                                    # loop through each definition
+                                    for m in range(0, len(definitions)):
+                                        # if the match == the definition key
+                                        if matches[y] == definitions[m][0]:
+                                            # substitute the definition value for the match
+                                            if type(definitions[m][1]) is str:
+                                                definitions[x][1] = definitions[x][1].replace('${' + matches[y] + '}', definitions[m][1])
```

Comment:
This is... SUPER deep. This needs to be refactored, but we can take care of that in another pass